geps/gep-3440/index.md

```diff
@@ -0,0 +1,281 @@
+trigger retries if specified, while errors like `403` and `404` are not retryable by default.
+
+### HAProxy
+1. **Retry Conditions**: HAProxy can retry requests based on various network conditions
+(e.g., connection failures, timeouts) and some HTTP error codes. While HAProxy does support gRPC via HTTP/2, it does not
+have built-in support for handling specific gRPC error codes (like `Cancelled`, `Deadline Exceeded`).
+It relies on HTTP-level conditions for retries, so its gRPC support is less granular than the GEP requires.
+2. **Retry Limits**: HAProxy allows you to set a maximum number of retries for a request using the `retries` directive.
+It also supports setting a timeout for the entire retry process using the `timeout connect` and `timeout server` directives.
+
+### Traefik
+1. **Retry Conditions**: Traefik allows for retries based on HTTP-level conditions (e.g., connection errors and
+certain HTTP status codes like 500, 502, 503, and 504), but it does not natively interpret specific gRPC error codes
+like `UNAVAILABLE` or `DEADLINE_EXCEEDED`. This means that, while Traefik can retry requests on common HTTP errors
+that might represent temporary issues, it lacks the ability to directly handle and retry based on
+gRPC-specific error codes, limiting its alignment with the GEP’s requirement for granular gRPC error handling.
+2. **Retry Limits**: Traefik provides configurable retry attempts and can set a maximum number of retries. However,
+Traefik does not offer per-try timeout controls specific to each retry attempt. Instead, it typically relies on a
+global request timeout, limiting the flexibility needed for more precise gRPC retry management (like Envoy’s `per_try_timeout`).
+
+## API
+Having a dedicated API for gRPC retry conditions is necessary because gRPC uses
+unique error codes (e.g., `UNAVAILABLE`, `DEADLINE_EXCEEDED`) that represent transient issues specific to its protocol,
+which are not adequately covered by general HTTP status codes. gRPC also supports streaming and real-time communications,
+making retry strategies more complex than those used for standard HTTP requests. Existing proxies like Envoy handle
+gRPC retries with specialized logic, while other proxies rely on HTTP error codes, lacking the precision needed
+for gRPC.
+
+### Go
+
+```go
+type GRPCRouteRule struct {
+    // Retry defines the configuration for when to retry a gRPC request.
+    //
+    // Support: Extended
+    //
+    // +optional
+    // <gateway:experimental>
+    Retry *GRPCRouteRetry `json:"retry,omitempty"`
+
+    // ...
+}
+
+// GRPCRouteRetry defines retry configuration for a GRPCRoute.
+//
+// Implementations SHOULD retry on common transient gRPC errors
+// if a retry configuration is specified.
+//
+type GRPCRouteRetry struct {
+    // Reasons defines the gRPC error conditions for which a backend request
+    // should be retried.
+    //
+    // Supported gRPC error conditions:
+    // * "cancelled"
```

Comment:
I dont have a great answer here but generally the string value here seems like a source of confusion. Why would it be `deadline-exceeded` and not `DEADLINE_EXCEEDED`? https://github.com/grpc/grpc-go/blob/a3a8657078983be4c5e9e9e856c757a5de9b3a45/codes/code_string.go#L72-L111

Reply:
> It seems that the API and the string values are designed based on what Envoy does but I am not sure if we want to follow Envoy's paradigm instead of following the GRPC-Core convention and require the canonical name.

Thanks for bringing this up. I didn't pay attention to this specific point. I wasn't initially trying to make the API fit the Envoy model. I think it is a fair callout, and we should mirror the gRPC spec rather than an impel motion. Will chang that!